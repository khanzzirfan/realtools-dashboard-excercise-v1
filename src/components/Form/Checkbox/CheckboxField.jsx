import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import Text from 'components/Text';

const StyledBox = styled.span`
  position: relative;
  display: inline-block;
  width: 14px;
  height: 14px;
  vertical-align: middle;
  background-color: var(--colors-linkWater);
  border: 1px solid var(--colors-spindle);
  border-radius: 2px;
  transition: all 150ms;

  &:hover {
    border: 1px solid var(--colors-irisBlue);
    box-sizing: border-box;
  }
`;

const StyledRectCheckbox = styled.span`
  display: block;
  position: absolute;
  top: 2px;
  left: 2px;
  width: 8px;
  height: 8px;
  background-color: var(--colors-irisBlue);
`;

const checkMarkStyle = css`
  border: 1px solid var(--colors-irisBlue);
  box-sizing: border-box;
`;

export const CheckboxContainer = styled.label`
  display: inline-flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 13px 8px 12px;
  border: 1.5px solid var(--colors-spindle);
  border-radius: 12px;
  outline: none;
  &:hover,
  &:active {
    background: var(--colors-linkWater);
    ${StyledBox} {
      ${checkMarkStyle}
    }
  }

  &:focus,
  &:focus-within {
    background: var(--colors-linkWater);
    box-shadow: 0px 0px 0px 2px var(--colors-irisBlue);
    ${StyledBox} {
      ${checkMarkStyle}
    }
  }
`;

// Hide checkbox visually but remain accessible to screen readers.
// Source: https://polished.js.org/docs/#hidevisually
const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  box-sizing: border-box;
`;

const Checkbox = ({ id, label, checked, name, ...props }) => (
  <React.Fragment>
    <CheckboxContainer
      htmlFor={`checkbox-${id || label}`}
      name={name}
      data-testid="checkboxcontainer"
    >
      <Text as="span" size="bodySmallMedium" mr={11} data-testid="checkboxlabel">
        {label}
      </Text>
      <HiddenCheckbox
        id={`checkbox-${id || label}`}
        checked={!!checked}
        value={name}
        name={name}
        {...props}
      />
      <StyledBox>{checked && <StyledRectCheckbox data-testid="innercheckmark" />}</StyledBox>
    </CheckboxContainer>
  </React.Fragment>
);

export default Checkbox;

Checkbox.propTypes = {
  checked: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  name: PropTypes.string.isRequired,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
};
