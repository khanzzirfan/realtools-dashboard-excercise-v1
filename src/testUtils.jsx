import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';
import { AppContextProvider } from './hooks/useAppContext';
import { theme } from './theme';

const AllTheProviders = ({ children }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        // âœ… turns retries off
        retry: 2,
      },
    },
  });
  return (
    <QueryClientProvider client={queryClient}>
      <AppContextProvider>
        <ThemeProvider theme={theme}>
          <MemoryRouter initialEntries={['/people', '/people/new', '/people/edit/:id']}>
            {children}
          </MemoryRouter>
        </ThemeProvider>
      </AppContextProvider>
    </QueryClientProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
