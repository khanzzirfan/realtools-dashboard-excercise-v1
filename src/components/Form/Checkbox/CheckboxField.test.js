import '@testing-library/jest-dom/extend-expect';
import { axe } from 'jest-axe';
import { render, screen, fireEvent } from '../../../testUtils';
import CheckboxField from './CheckboxField';

describe('Checkbox', () => {
  it('should renders correctly with text contractor and checked false', () => {
    const onChangeFn = jest.fn();
    render(
      <CheckboxField label="Contractor" name="contractor" checked={false} onChange={onChangeFn} />
    );
    const element = screen.getByTestId('checkboxlabel');
    expect(element).toHaveTextContent('Contractor');
    // not.toBeInTheDocument
    expect(screen.queryAllByTestId('innercheckmark')).toEqual([]);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox.checked).toEqual(false);
  });

  it('should renders correctly with text contractor and checked state true', () => {
    const onChangeFn = jest.fn();
    render(
      <CheckboxField label="Contractor" name="contractor" checked={true} onChange={onChangeFn} />
    );
    const element = screen.getByTestId('checkboxlabel');
    expect(element).toHaveTextContent('Contractor');
    // not.toBeInTheDocument
    expect(screen.queryByTestId('innercheckmark')).toBeInTheDocument();
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox.checked).toEqual(true);
  });

  it('should trigger onChange callback when user selected', () => {
    const onChangeFn = jest.fn();
    render(
      <CheckboxField label="Contractor" name="contractor" checked={true} onChange={onChangeFn} />
    );
    fireEvent.click(screen.getByLabelText(/Contractor/i));
    expect(onChangeFn).toHaveBeenCalledTimes(1);
  });

  it('should render checkbox with default styles', () => {
    const onChangeFn = jest.fn();
    render(
      <CheckboxField label="Contractor" name="contractor" checked={true} onChange={onChangeFn} />
    );
    const node = screen.getByTestId('checkboxcontainer');
    expect(node).toMatchSnapshot();
  });

  /** axe voilations check */
  it('should not have axe voilations', async () => {
    const onChangeFn = jest.fn();
    const { container } = render(
      <CheckboxField label="Contractor" name="contractor" checked={true} onChange={onChangeFn} />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
