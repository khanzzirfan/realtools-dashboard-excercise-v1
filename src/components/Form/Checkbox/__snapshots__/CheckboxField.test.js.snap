// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render checkbox with default styles 1`] = `
.c1 {
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.58;
  margin: 0;
  margin-right: 11px;
}

.c4 {
  position: relative;
  display: inline-block;
  width: 14px;
  height: 14px;
  vertical-align: middle;
  background-color: var(--colors-linkWater);
  border: 1px solid var(--colors-spindle);
  border-radius: 2px;
  -webkit-transition: all 150ms;
  transition: all 150ms;
}

.c4:hover {
  border: 1px solid var(--colors-irisBlue);
  box-sizing: border-box;
}

.c5 {
  display: block;
  position: absolute;
  top: 2px;
  left: 2px;
  width: 8px;
  height: 8px;
  background-color: var(--colors-irisBlue);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 13px 8px 12px;
  border: 1.5px solid var(--colors-spindle);
  border-radius: 12px;
  outline: none;
}

.c0:hover,
.c0:active {
  background: var(--colors-linkWater);
}

.c0:hover .c3,
.c0:active .c3 {
  border: 1px solid var(--colors-irisBlue);
  box-sizing: border-box;
}

.c0:focus,
.c0:focus-within {
  background: var(--colors-linkWater);
  box-shadow: 0px 0px 0px 2px var(--colors-irisBlue);
}

.c0:focus .c3,
.c0:focus-within .c3 {
  border: 1px solid var(--colors-irisBlue);
  box-sizing: border-box;
}

.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  box-sizing: border-box;
}

<label
  class="c0"
  data-testid="checkboxcontainer"
  for="checkbox-Contractor"
  name="contractor"
>
  <span
    class="c1"
    data-testid="checkboxlabel"
  >
    Contractor
  </span>
  <input
    checked=""
    class="c2"
    id="checkbox-Contractor"
    name="contractor"
    type="checkbox"
    value="contractor"
  />
  <span
    class="c3 c4"
  >
    <span
      class="c5"
      data-testid="innercheckmark"
    />
  </span>
</label>
`;
